text/0000-cargo-workspace.md

```diff
@@ -0,0 +1,348 @@
+*either* specify `package.workspace` or specify `[workspace]`. That is, a
+crate cannot both be a root in a workspace (contain `[workspace]`) and also be
+a member of another workspace (contain `package.workspace`).
+
+### "Virtual" `Cargo.toml`
+
+A good number of projects do not necessarily have a "root `Cargo.toml`" which is
+an appropriate root for a workspace. To accommodate these projects and allow for
+the output of a workspace to be configured regardless of where crates are
+located, Cargo will now allow for "virtual manifest" files. These manifests will
+currently **only** contains the `[workspace]` table and will notably be lacking
+a `[project]` or `[package]` top level key.
+
+Cargo will for the time being disallow many commands against a virtual manifest,
+for example `cargo build` will be rejected. Arguments that take a package,
+however, such as `cargo test -p foo` will be allowed. Workspaces can eventually
+get extended with `--all` flags so in a workspace root you could execute
+`cargo build --all` to compile all crates.
+
+### Validating a workspace
+
+A workspace is valid if these two properties hold:
+
+1. A workspace has only one root crate (that with `[workspace]` in
+   `Cargo.toml`).
+2. All workspace crates defined in `workspace.members` point back to the
+   workspace root with `package.workspace`.
+
+While the restriction of one-root-per workspace may make sense, the restriction
+of crates pointing back to the root may not. If, however, this restriction were
+not in place then the set of crates in a workspace may differ depending on
+which crate it was viewed from. For example if workspace root A includes B then
+it will think B is in A's workspace. If, however, B does not point back to A,
+then B would not think that A was in its workspace. This would in turn cause the
+set of crates in each workspace to be different, further causing `Cargo.lock` to
+get out of sync if it were allowed. By ensuring that all crates have edges to
+each other in a workspace Cargo can prevent this situation and guarantee robust
+builds no matter where they're executed in the workspace.
+
+To alleviate misconfiguration Cargo will emit an error if the two properties
+above do not hold for any crate attempting to be part of a workspace. For
+example, if the `package.workspace` key is specified, but the crate is not a
+workspace root or doesn't point back to the original crate an error is emitted.
+
+### Implicit relations
+
+The combination of the `package.workspace` key and `[workspace]` table is enough
+to specify any workspace in Cargo. Having to annotate all crates with a
+`package.workspace` parent or a `workspace.members` list can get quite tedious,
+however! To alleviate this configuration burden Cargo will allow these keys to
+be implicitly defined in some situations.
+
+The `package.workspace` can be omitted if it would only contain `../` (or some
+repetition of it). That is, if the root of a workspace is hierarchically the
```

Comment:
Actually I found "hierarchically the first" confusing; did it mean the closest to the root directory of the system (`/`) or the furthest from it?

I interpreted it as "the furthest" (because it would be the "first" to be found by a repeated application of `../`); then the root of the workspace `ws0` is *not* "the first" `Cargo.toml` with `[workspace]` regardless of `ws0/ws1`'s (lack of) dependency on `ws0/ws1/util`.

My question could be rephrased as: can an unrelated `Cargo.toml` (`ws0/ws1/Cargo.toml` in this case) break the validity of a workspace (`ws0`) that depends on implicit `package.workspace` ?